### PR TITLE
NVMe support in PBA

### PR DESCRIPTION
**Pull request by verglor**
[PR#1](https://github.com/sedutil/sedutil/pull/1) with fixed formatting
Original [PR#108](https://github.com/Drive-Trust-Alliance/sedutil/pull/108) + https://github.com/jojolepirate/sedutil/commit/c91eb0cafeded2263d5090ab273a16f859674d11 + [PR#56](https://github.com/Drive-Trust-Alliance/sedutil/pull/56)

### CODE REVIEW

**Review by mflav — COMMENTED** [review]
I do not believe this will compile as intended. The define directive needs to be split up using backslash. See my update to the original commit.

File: LinuxPBA/UnlockSEDs.cpp [hunk 1]
```diff
@@ -35,7 +35,11 @@ uint8_t UnlockSEDs(char * password) {
     LOG(D4) << "Enter UnlockSEDs";
     mvprintw(6,2,"Scanning....");
     while (TRUE) {
-        snprintf(devref,23,"/dev/sd%c",(char) 0x61+i);
+        switch(DtaDevOS::getNextDevice(i)){
+            case 1  : snprintf(devref,23,"/dev/%s",DtaDevOS::getDeviceName()); break;
+            case -1 : snprintf(devref,23,"/dev/sd%c",(char) 0x61+i);    break;
```
Comment (cristim) [inline LinuxPBA/UnlockSEDs.cpp]: again use 'a'+i here to make it more obvious

File: linux/DtaDevLinuxNvme.h [hunk 2]
```diff
@@ -18,7 +18,13 @@ along with sedutil.  If not, see <http://www.gnu.org/licenses/>.
 
  * C:E********************************************************************** */
 #pragma once
-#include "linux/nvme.h"
+#include <linux/version.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 4, 0)
```
Comment (cristim) [inline linux/DtaDevLinuxNvme.h:22]: I propose to just update the kernel to 4.12, perhaps we can then stop distributing this patch.

Would that work?

File: linux/os.h [hunk 3]
```diff
@@ -31,5 +31,13 @@ along with sedutil.  If not, see <http://www.gnu.org/licenses/>.
 #define FALSE 0
 // a few OS specific methods that need to be worked out
 #define SNPRINTF snprintf
-#define DEVICEMASK snprintf(devname,23,"/dev/sd%c",(char) 0x61+i)
+// Cumbersome, but trying to avoid editing common files. If /sys/block is not present, reverts to enumerating /dev/sd* devices. 
+#define DEVICEMASK switch(DtaDevOS::getNextDevice(i)) {
+    case 1 : snprintf(devname,23,"/dev/%s",DtaDevOS::getDeviceName()); 
+        break;
+    case -1 : snprintf(devname,23,"/dev/sd%c",(char) 0x61+i);
```
Comment (cristim) [inline linux/os.h]: again 'a'+i

**Comment by cristim** [comment]
@laser-beans which line exactly are you referring to?

File: LinuxPBA/UnlockSEDs.cpp (hunk 1 above)
Comment (verglor) [reply LinuxPBA/UnlockSEDs.cpp]: I believe it is not related to this PR however I have pushed this refactoring.

File: linux/DtaDevLinuxNvme.h (hunk 2 above)
Comment (verglor) [reply linux/DtaDevLinuxNvme.h:22]: I really don't know. I just put together all NVMe related commits from [jojolepirate repo](https://github.com/jojolepirate/sedutil/).

**Review by cristim — APPROVED** [review]

File: linux/DtaDevLinuxNvme.h (hunk 2 above)
Comment (cristim) [reply linux/DtaDevLinuxNvme.h:22]: Fair enough, let's do this one step at a time.

**Pull request merged**